### DISCUSSION TITLE
[Typescript] Actions firing after conditions

### DISCUSSION THREAD

**Discussion by EightArmCode** [statelyai/xstate#3433] [Q&A] [answered]
After a few days of trying to figure out why an assign action wasn't firing, I figured out that the guard was firing first. Since the action contains the needed context to evaluate the condition, the condition always failed.

I have searched and found a few issues which indicate there's some work underway in regard to sequencing. Now I understand why!

My question is simple - is there a way to make the action fire before the condition? In my case, I'm invoking a promise, and the actions are inside the `onDone`, so this was particularly puzzling! I expected the promise would be resolved by the time it enters `onDone`. The irony is that the `event` parameter being passed to the `guard` contains the payload, but it seems proper to first assign the response to the context and evaluate the condition from machine context. I would also prefer not to assign an arbitrary delay interval if it can be avoided.

Here's a working [sandbox](https://codesandbox.io/s/thirsty-margulis-8k1471?file=/src/getDownload.ts). If you open the console, you'll see the sequencing, and indeed, the console logs inside the assign action are invoked _after_ the conditions are evaluated.

_Note about the code sandbox, it will error if you try to refresh the UI window. Just refresh the whole page._

**Comment by Andarist** [comment]
Could you clearly describe which action and guard this is related to? An expected and actual notes would very useful - you have shared a somewhat generic description of the problem with a concrete example and I'm unsure how to relate one to another right now.

**Reply by EightArmCode** [reply]
Ah yes, sorry, that'd be helpful, wouldn't it?

The `assignDownloadResponse` is the action that I was intending to complete prior to the `cond` prop in the `getDownload.onDone` object.

**Reply by Andarist** [reply]
Yeah, so your solution and David's solution are both fine - it just depends on the personal preference. I personally wouldn't mind "duplicating" the `assignDownloadResponse` in both "branches" of the `onDone`.

It's interesting though that SCXML has a way to do what you want - it's available as [`<finalize/>`](https://www.w3.org/TR/scxml/#finalize), cc @davidkpiano . I'm not sure if the `finalize` keyword is really the best name for this, but perhaps we could explore this or alternatives in v5.

**Reply by EightArmCode** [reply]
Indeed there are several different ways to solve this. I was second guessing my decision not to pull the payload out of the event when checking the condition.

**Reply by EightArmCode** [reply]
Thinking more about the finalize option. It sounds pretty good, actually. I have the impression that it would provide a way to assign the data at the end of the service invocation. Is that right?

**Reply by Andarist** [reply]
> <finalize> contains executable content that is executed whenever the external service returns an event after the <invoke> has been executed.

It's not only for handling "data at the end of the service invocation". It can be used to handle any incoming event from the given `<invoke/>`. This part of the spec mentions that it can contain "executable content" and that indicates that it could handle more than just `assign`s (even just functions, just like your entry actions etc). However, the SCXML test suite only uses `assign`s in it.

**Comment by davidkpiano** [comment, marked as answer]
You can use a "transient state" (a state that immediately transitions to another state) to accomplish this:

```js
        getDownload: {
          invoke: {
            id: "getDownload",
            src: "getDownload",
            // onDone: [
            //   {
            //     target: "enterPassword",
            //     cond: "requiresPassword",
            //     actions: "assignDownloadResponse"
            //   },
            //   {
            //     target: "downloadReady",
            //     actions: "assignDownloadResponse"
            //   }
            // ],
            onDone: 'determine',
            onError: {
              target: "error",
              actions: "assignDownloadError"
            }
          }
        },
        determine: {
          entry: 'assignDownloadResponse',
          always: [
            {
              target: "enterPassword",
              cond: "requiresPassword"
            },
            {
              target: "downloadReady"
            }
          ]
        },
```

**Reply by EightArmCode** [reply]
Is it expected/intentional that `onDone` runs before the invoked promise is resolved? If so, can you help me understand that?

**Reply by EightArmCode** [reply]
Or does the condition simply fire before the action? I suppose now that I think about it again, that the condition event contains the payload, so it's just about the sequence.

**Reply by davidkpiano** [reply]
The condition is used to determine if the transition is _taken_. A transition will not be taken if the condition is false, and its actions will not be exeucted.

**Reply by Andarist** [reply]
> Is it expected/intentional that onDone runs before the invoked promise is resolved? If so, can you help me understand that?

Nope, the promise has to resolve first for the `onDone` to get checked and potentially "executed".

**Reply by EightArmCode** [reply]
Yeah, I feel a bit silly now, tbh. My own condition was blocking me.

**Comment by EightArmCode** [comment]
Thanks for the quick feedback! I'm finally getting the hang of it.

While I have your attention, I was wondering about something. Somewhere in the last few days I came across a post where it was said that it's no longer necessary to do `(event as any).data`, but I couldn't find my way back to it. Does that ring any bells for either of you?

**Reply by Andarist** [reply]
Have you tried our typegen? You should be able to find the docs for it in our docs

**Reply by EightArmCode** [reply]
Ah yes, it's currently broken because we aren't passing a literal object to createMachine. I am inspired to create more submachines in place of compound states, but it will take me some time to write and test that, and our focus has changed direction for the moment.
